### PR TITLE
[MetaSchedule][UX] Make `Database` with-able

### PR DESCRIPTION
**Pull request by junrushao**
`ApplyHistoryBest` right now plays a role as the database adaptor to query inside the database. In fact, the logic could be simplified and users only have to deal with `Database` instead of this extra object.

- [x] Add `EnterWithScope`/`ExitWithScope`/`Current` to Database
- [x] Migrate `te_filter_func` => "tir_filter" in Relay's pass context
- [x] Migrate `f_take_tuning_record` => "Database.query_tuning_record"
- [x] Migrate `TECompiler` to use `Database` 
- [x] Remove apply-history-best

Next PR:
- Migrate `f_direct_dispatch` (potentially unify with `apply_fixed_schedule`?)

cc @Hzfengsy @junrushao1994

### CODE REVIEW

**Comment by junrushao** [comment]
CC: @masahi @Hzfengsy @spectrometerHBH @vinx13

**Comment by junrushao** [comment]
Also CC @sunggg @YuchenJin for updates on Relax side

**Review by MasterJH5574 — COMMENTED** [review]
Some tiny nits :-)

**Review by MasterJH5574 — COMMENTED** [review]
Another problem I’m thinking about is.. What will happen if we have code
```python
with database_0, database_1:
    ...
```

If we query the best record in the “with” body, looks like the current impl will only looking for records in `database_2`, rather than looking for record from both databases?

**Comment by masahi** [comment]
> Another problem I’m thinking about is.. What will happen if we have code
> 
> ```python
> with database_0, database_1:
>     ...
> ```
> 
> If we query the best record in the “with” body, looks like the current impl will only looking for records in `database_2`, rather than looking for record from both databases?


I think if we have a way to merge several databases, we can agree that we only support `with database:`.

**Review by masahi — APPROVED** [review]
I really like this change! Modulo the issue raised by @MasterJH5574, LGTM.

File: include/tvm/meta_schedule/database.h [hunk 1]
```diff
@@ -203,6 +203,27 @@ class DatabaseNode : public runtime::Object {
...
   virtual int64_t Size() = 0;
+  /*!
+   * \brief Query the best record of the given workload from the database.
+   * \param mod The IRModule to be searched for.
+   * \param target The target to be searched for.
+   * \return The best record of the given workload; NullOpt if not found.
+   */
+  virtual Optional<TuningRecord> QueryTuningRecord(IRModule mod, Target target);
+  /*!
+   * \brief Query the best record of the given workload from the database.
```
Comment (zxybazh) [inline include/tvm/meta_schedule/database.h:214]: Ditto, `record` -> `Schedule`

File: src/meta_schedule/database/database.cc [hunk 2]
```diff
@@ -154,6 +154,59 @@ TuningRecord TuningRecord::FromJSON(const ObjectRef& json_obj, const Workload& w
   return TuningRecord(trace, workload, run_secs, target, args_info);
 }
 
+/******** Database ********/
+
+Optional<TuningRecord> DatabaseNode::QueryTuningRecord(IRModule mod, Target target) {
```
Comment (zxybazh) [inline src/meta_schedule/database/database.cc:159]: We may want to add a target comparison function to distinguish between the target given here in the future.

**Review by zxybazh — APPROVED** [review]
Generally LGTM and we may want to add function to distinguish between targets, which should be a minor problem for now.

File: src/relay/backend/utils.cc [hunk 3]
```diff
@@ -368,6 +371,71 @@ void BindParamsInModule(IRModule mod, Map<String, runtime::NDArray> params) {
   BindParamsInModule(mod, params_tmp);
 }
 
+Optional<tir::PrimFunc> DefaultTIRConverterImpl(const Array<te::Tensor>& args,
```
Comment (sunggg) [inline src/relay/backend/utils.cc:374]: Add some comments what this function does.

**Review by sunggg — COMMENTED** [review]
Thank you for the great work! 
Overall, LGTM. A few questions and nits.

File: python/tvm/meta_schedule/database/database.py [hunk 4]
```diff
@@ -234,6 +234,71 @@ def __len__(self) -> int:
         """
         return _ffi_api.DatabaseSize(self)  # type: ignore # pylint: disable=no-member
 
+    def query_tuning_record(self, mod: IRModule, target: Target) -> Optional[TuningRecord]:
```
[the remark this answers is not in the document]
Comment (zxybazh) [reply python/tvm/meta_schedule/database/database.py:237]: I think we can access them with `get_top_k` function instead.

File: include/tvm/meta_schedule/database.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (junrushao) [reply include/tvm/meta_schedule/database.h:214]: thanks for spotting this!

File: src/meta_schedule/database/database.cc (hunk 2 above)
Comment (junrushao) [reply src/meta_schedule/database/database.cc:159]: that makes sense!

File: python/tvm/meta_schedule/relay_integration.py [hunk 5]
```diff
@@ -38,7 +36,7 @@ def extract_task_from_relay(
     opt_level: int = 3,
     pass_config: Optional[Dict[str, Any]] = None,
     disabled_pass: Optional[List[str]] = None,
-    te_filter_func: Union[str, None, Callable[[List[Tensor]], PrimFunc]] = None,
+    tir_converter: str = "default",
```
[the remark this answers is not in the document]
Comment (junrushao) [reply python/tvm/meta_schedule/relay_integration.py:39]: Oh looks like I missed the doc here! Added a few lines to explain

File: python/tvm/meta_schedule/database/database.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (junrushao) [reply python/tvm/meta_schedule/database/database.py:237]: Yes we expose `get_top_k` and `get_all_tuning_records` APIs, but it's supposed to be less used than `query_*`

File: src/relay/backend/te_compiler_cache.cc [hunk 6]
```diff
@@ -359,32 +350,43 @@ class ScheduleBuilder : public ExprVisitor {
...
+        if (Optional<PrimFunc> f = tir_converter(te_args, constants)) {
+          if (Optional<TuningRecord> opt_record = database_.value()->QueryTuningRecord(
+                  /*mod=*/backend::PrimFuncToIRModule(f.value()),
+                  /*target=*/target_)) {
+            static InstructionKind kind_transform_layout = InstructionKind::Get("TransformLayout");
+            TuningRecord record = opt_record.value();
+            for (const Instruction& inst : record->trace->insts) {
+              if (inst->kind.same_as(kind_transform_layout)) {
+                ICHECK_EQ(inst->attrs.size(), 3);
+                MetaScheduleLayoutRewriter::LayoutQueuePush(Downcast<IndexMap>(inst->attrs[2]));
```
[the remark this answers is not in the document]
Comment (junrushao) [reply src/relay/backend/te_compiler_cache.cc:376]: Oh this is an extremely hacky version of weight layout rewriting on CPU

File: src/relay/backend/te_compiler_cache.cc [hunk 7]
```diff
@@ -359,32 +350,43 @@ class ScheduleBuilder : public ExprVisitor {
...
-                  /*mod=*/relay_mod,                        //
-                  /*target=*/target_,                       //
-                  /*dispatched=*/Array<IRModule>{tir_mod},  //
-                  /*f_take_tuning_record=*/ExtractTransformLayout)) {
-            IRModule scheduled_mod =
-                tir::transform::RemoveWeightLayoutRewriteBlock()(opt_scheduled_mod.value());
-            ICHECK_EQ(scheduled_mod->functions.count(prim_fn_var), 1);
-            prim_func = Downcast<tir::PrimFunc>(scheduled_mod->functions[prim_fn_var]);
+        if (Optional<PrimFunc> f = tir_converter(te_args, constants)) {
+          if (Optional<TuningRecord> opt_record = database_.value()->QueryTuningRecord(
```
[the remark this answers is not in the document]
Comment (junrushao) [reply src/relay/backend/te_compiler_cache.cc:368]: In TECompiler, it falls back to default schedules defined in TOPI directly. We want to support joined database as Masa suggested [here](https://github.com/apache/tvm/pull/12520#issuecomment-1228888860) in the near future for falling back to other databases

File: src/relay/backend/utils.cc (hunk 3 above)
Comment (junrushao) [reply src/relay/backend/utils.cc:374]: added!

**Comment by junrushao** [comment]
> > Another problem I’m thinking about is.. What will happen if we have code
> > ```python
> > with database_0, database_1:
> >     ...
> > ```
> >   
> > If we query the best record in the “with” body, looks like the current impl will only looking for records in `database_2`, rather than looking for record from both databases?
> 
> I think if we have a way to merge several databases, we can agree that we only support `with database:` (one database).

@MasterJH5574 @masahi Great discussion here! We will introduced `JoinedDatabase` in the near future to cater such needs

**Comment by MasterJH5574** [comment]
> > > Another problem I’m thinking about is.. What will happen if we have code
> > > ```python
> > > with database_0, database_1:
> > >     ...
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > If we query the best record in the “with” body, looks like the current impl will only looking for records in `database_2`, rather than looking for record from both databases?
> > 
> > 
> > I think if we have a way to merge several databases, we can agree that we only support `with database:` (one database).
> 
> @MasterJH5574 @masahi Great discussion here! We will introduced `JoinedDatabase` in the near future to cater such needs

@junrushao Sounds a great idea. Thanks in advance!

**Comment by tqchen** [comment]
One minor nit on naming :) Join have a separate meaning and in this case perhaps concat(or merge) is more appropriate. Try lookup the terminology in pandas

**Comment by junrushao** [comment]
@tqchen Sure `join` doesn't look like a proper name. Perhaps `ms.database.MergedDatabase` sounds better?

**Pull request merged**